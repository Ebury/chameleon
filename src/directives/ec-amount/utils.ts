
import defaultOptions from './options';
import type { AmountDirectiveOptions } from './types';

function format(input: undefined, opts?: AmountDirectiveOptions): undefined;
function format(input: null, opts?: AmountDirectiveOptions): null;
function format(input: string | number, opts?: AmountDirectiveOptions): string;
function format(input: string | number | undefined | null, opts: AmountDirectiveOptions = defaultOptions): string | null | undefined {
  /* eslint-disable no-param-reassign */
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.startsWith('-') ? '-' : '';

  input = sanitizeInput(input, opts.decimalSeparator);
  if (opts.precision > 0) {
    input = keepOnlyOneDecimalSeparator(input, opts.decimalSeparator);
  } else {
    input = removeDecimal(input, opts.decimalSeparator);
  }
  const parts = input.split(opts.decimalSeparator);
  let integralPart = parseToStr(parts[0]);
  let fractionalPart = parts[1];

  if (fractionalPart && fractionalPart.length > opts.precision) {
    fractionalPart = fractionalPart.substring(0, opts.precision);
  }

  integralPart = addGroupingSeparators(integralPart, opts.groupingSeparator);
  const result = negative + joinIntegralAndFractionalParts(integralPart, fractionalPart, opts.decimalSeparator);
  return result;
}

function parseToStr(number: string): string {
  const parsed = Number.parseFloat(number);
  if (Number.isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number: string, decimalSeparator: string): string {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function removeDecimal(number: string, decimalSeparator: string): string {
  const split = number.split(decimalSeparator);
  return split[0];
}

function unFormat(input: string, groupingSeparator: string, decimalSeparator: string): string {
  if (!input) {
    return input;
  }

  return input
    .replace(new RegExp(`\\${groupingSeparator}`, 'g'), '')
    .replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
}

function sanitizeInput(input: string, decimalSeparator: string): string {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integralPart: string, separator: string): string {
  return integralPart.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegralAndFractionalParts(integralPart: string, fractionalPart: string, decimalSeparator: string): string {
  return fractionalPart !== undefined ? integralPart + decimalSeparator + fractionalPart : integralPart;
}

/* c8 ignore start */
function setCursor(el: HTMLInputElement, position: number) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}
/* c8 ignore stop */

export {
  format,
  sanitizeInput,
  setCursor,
  unFormat,
};
