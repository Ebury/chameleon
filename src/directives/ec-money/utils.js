import defaults from './options';

function format(input, opt = defaults) {
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.includes('-') ? '-' : '';

  input = sanitizeInput(input, opt.decimalSeparator);
  input = keepOnlyOneDecimalSeparator(input, opt.decimalSeparator);
  const parts = input.split(opt.decimalSeparator);
  let integer = parseToStr(parts[0]);
  let decimal = parts[1];

  if (decimal && decimal.length > opt.precision) {
    decimal = decimal.substr(0, opt.precision);
  }

  integer = addGroupingSeparators(integer, opt.groupingSeparator);
  const result = negative + joinIntegerAndDecimal(integer, decimal, opt.decimalSeparator);
  return result;
}

function parseToStr(number) {
  /* eslint no-restricted-globals: "off" */
  const parsed = parseFloat(number);
  if (isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number, decimalSeparator) {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function unFormat(input, groupingSeparator, decimalSeparator) {
  /* eslint no-param-reassign: "off" */
  input = (`${input}`).replace(new RegExp(`\\${groupingSeparator}`, 'g'), '');
  input = (`${input}`).replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
  return input;
}

function sanitizeInput(input, decimalSeparator) {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integer, separator) {
  return integer.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegerAndDecimal(integer, decimal, decimalSeparator) {
  return decimal !== undefined ? integer + decimalSeparator + decimal : integer;
}

function setCursor(el, position) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}

// https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events#The_old-fashioned_way
function event(name) {
  const evt = document.createEvent('Event');
  evt.initEvent(name, true, true);
  return evt;
}

export {
  format,
  unFormat,
  setCursor,
  event,
  sanitizeInput,
};
