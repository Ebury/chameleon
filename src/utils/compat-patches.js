/* istanbul ignore file */

import {
  Dropdown as FvDropdown,
  TooltipDirective as FvTooltipDirective,
  VClosePopper,
  VTooltip,
} from 'floating-vue';

// we need to tell the @vue/compat that all components coming from floating-vue are Vue 3 components so it skips
// them when patching component instances. These components can't run in default mode for Vue 2, there is lots of
// errors coming from the console in that mode.
export default function applyCompatPatches() {
  const compatConfig = {
    MODE: 3,
    RENDER_FUNCTION: false,
  };

  VTooltip.compatConfig = compatConfig;
  VClosePopper.compatConfig = compatConfig;
  FvTooltipDirective.compatConfig = compatConfig;
  for (const component of Object.values(FvTooltipDirective.components)) {
    component.compatConfig = compatConfig;
  }

  FvDropdown.compatConfig = compatConfig;

  for (const component of Object.values(FvDropdown.components)) {
    component.compatConfig = compatConfig;
  }
}
