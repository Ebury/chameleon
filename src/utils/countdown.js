import mitt from 'mitt';

export default class Countdown {
  constructor() {
    const m = mitt();
    Object.assign(this, m);
  }

  reduceMinutesLeft(totalSeconds) {
    this.getminutes(totalSeconds);
    this.getseconds(totalSeconds, this.minutesLeft);
  }

  start(secondsToGo, showMinutes) {
    this.seconds = secondsToGo;
    this.secondsLeft = secondsToGo;
    this.currentTime = Math.ceil(Date.now() / 1000);
    this.startTime = Math.ceil(Date.now() / 1000);
    this.interval = setInterval(() => this.reduceSecondsLeft(showMinutes), 1000);
  }

  get timeDifference() {
    return this.currentTime - this.startTime;
  }

  reduceSecondsLeft(showMinutes) {
    this.currentTime = Math.ceil(Date.now() / 1000);
    this.totalSeconds = this.seconds - this.timeDifference;
    this.minutesLeft = null;
    this.secondsLeft = this.totalSeconds;

    if (showMinutes) {
      this.reduceMinutesLeft(this.totalSeconds);
    }

    this.emit('time-updated', {
      seconds: this.secondsLeft,
      minutes: this.minutesLeft,
      totalSeconds: this.totalSeconds,
    });

    if (this.totalSeconds <= 0) {
      this.stop();
      this.emit('time-expired');
    }
  }

  stop() {
    clearInterval(this.interval);
  }

  getminutes(totalSeconds) {
    this.minutesLeft = Math.floor(totalSeconds / 60);
  }

  getseconds(totalSeconds, minutes) {
    // take mins remaining (as seconds) away from total seconds remaining
    this.secondsLeft = totalSeconds - Math.round(minutes * 60);
  }
}
