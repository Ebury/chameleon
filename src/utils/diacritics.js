export function removeDiacritics(str) {
  if (typeof str === 'string' && str.normalize) {
    // https://stackoverflow.com/questions/990904/remove-accents-diacritics-in-a-string-in-javascript
    return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  }

  // IE11 users will not enjoy this feature:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize
  return str;
}
