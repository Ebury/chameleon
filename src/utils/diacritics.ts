export function removeDiacritics(str: string): string {
  if (str && str.normalize) {
    // https://stackoverflow.com/questions/990904/remove-accents-diacritics-in-a-string-in-javascript
    return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  }

  // IE11 users will not enjoy this feature:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize
  return str;
}
