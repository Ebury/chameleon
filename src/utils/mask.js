export function mask(text, maskSymbol = '*', visibleChars = 2) {
  if (!text) {
    throw new Error('Text is required');
  }
  if (!maskSymbol) {
    throw new Error('Mask symbol cannot be empty');
  }
  if (!visibleChars || visibleChars < 0) {
    throw new Error('Visible characters must be a number greater than zero');
  }

  const digitsToBeMasked = text.length - visibleChars;
  return maskSymbol.repeat(digitsToBeMasked) + text.substr(digitsToBeMasked);
}
