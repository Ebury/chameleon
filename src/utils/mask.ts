export function mask(text: string, maskSymbol: string = '*', visibleChars: number = 2): string {
  if (!text) {
    throw new Error('Text is required');
  }
  if (!maskSymbol) {
    throw new Error('Mask symbol cannot be empty');
  }
  if (!visibleChars || visibleChars < 0) {
    throw new Error('Visible characters must be a number greater than zero');
  }

  const digitsToBeMasked = text.length - visibleChars;
  return maskSymbol.repeat(digitsToBeMasked) + text.substring(digitsToBeMasked);
}
