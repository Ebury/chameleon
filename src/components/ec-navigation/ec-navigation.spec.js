import { mount } from '@vue/test-utils';

import { withMockedConsole } from '../../../tests/utils/console';
import EcNavigation from './ec-navigation.vue';

describe('EcNavigation', () => {
  function mountNavigation(opts, mountOpts) {
    return mount(EcNavigation, {
      props: {
        isCollapsed: false,
        isCollapsable: false,
        branding: {},
        ...opts,
      },
      ...mountOpts,
    });
  }

  it('should throw an error if required props are missing', () => {
    withMockedConsole((errorSpy, warnSpy) => {
      mount(EcNavigation);

      expect(warnSpy).toHaveBeenCalledTimes(1);
      expect(warnSpy.mock.calls[0][0]).toContain('Missing required prop: "isCollapsable"');
    });
  });

  it('should be expanded when isCollapsed is set to false', () => {
    const wrapper = mountNavigation({ isCollapsed: false });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.classes('ec-navigation--is-collapsed')).toBe(false);
  });

  it('should be collapsed when isCollapsed is set to true', () => {
    const wrapper = mountNavigation({ isCollapsed: true });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.classes('ec-navigation--is-collapsed')).toBe(true);
  });

  it('should be collapsable when isCollapsable is set', () => {
    const wrapper = mountNavigation({ isCollapsable: true });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.classes('ec-navigation--is-collapsable')).toBe(true);
  });

  it('should not render branding if no logo is given in branding object', () => {
    const wrapper = mountNavigation({ branding: { logo: null } });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findByDataTest('ec-navigation__branding').exists()).toBe(false);
  });

  it('should not render branding if logo is given in branding object but showBrandingLogo is set to false', () => {
    const wrapper = mountNavigation({
      branding: { logo: '/img/my.png' },
      showBrandingLogo: false,
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findByDataTest('ec-navigation__branding').exists()).toBe(false);
  });

  it('should render branding logo and name when given', () => {
    const wrapper = mountNavigation({
      branding: { logo: '/img/my.png', name: 'My Branding' },
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findByDataTest('ec-navigation__branding').element).toMatchSnapshot('Branding should have alt and src.');
  });

  it('should only render mandatory slot if no other slots were given', () => {
    const wrapper = mountNavigation();
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findAllByDataTest('ec-navigation__block').length).toBe(1);
    expect(wrapper.findByDataTest('ec-navigation__menu').exists()).toBe(true);
  });

  it('should render all given slots', () => {
    const wrapper = mountNavigation(null, {
      slots: {
        'user-info': '<div>User Info</div>',
        'call-to-action': '<div>CTA</div>',
        menu: '<div>Menu</div>',
        'footer-menu': '<div>Footer Menu</div>',
        copyright: '<div>Copyright</div>',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findAllByDataTest('ec-navigation__block').length).toBe(5);
  });
});
