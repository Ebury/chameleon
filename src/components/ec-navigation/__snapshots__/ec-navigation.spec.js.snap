// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcNavigation should be collapsible when canBeCollapsed is set 1`] = `
<div
  class="ec-navigation ec-navigation--can-be-collapsed"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`EcNavigation should be expanded when isExpanded is set 1`] = `
<div
  class="ec-navigation ec-navigation--is-expanded"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`EcNavigation should not render branding if logo is given in branding object but showBrandingLogo is set to false 1`] = `
<div
  class="ec-navigation"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`EcNavigation should not render branding if no logo is given in branding object 1`] = `
<div
  class="ec-navigation"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`EcNavigation should only render mandatory slot if no other slots were given 1`] = `
<div
  class="ec-navigation"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`EcNavigation should render all given slots 1`] = `
<div
  class="ec-navigation"
>
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__user-info"
  >
    <div>
      User Info
    </div>
  </div>
   
  <div
    class="ec-navigation__block ec-navigation__call-to-action"
  >
    <div>
      CTA
    </div>
  </div>
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  >
    <div>
      Menu
    </div>
  </div>
   
  <div
    class="ec-navigation__block ec-navigation__footer-menu"
  >
    <div>
      Footer Menu
    </div>
  </div>
   
  <div
    class="ec-navigation__block ec-navigation__copyright"
  >
    <div>
      Copyright
    </div>
  </div>
</div>
`;

exports[`EcNavigation should render branding logo and name when given 1`] = `
<div
  class="ec-navigation"
>
  <div
    class="ec-navigation__branding"
  >
    <img
      alt="My Branding"
      src="/img/my.png"
    />
  </div>
   
  <!---->
   
  <!---->
   
  <div
    class="ec-navigation__block ec-navigation__menu"
  />
   
  <!---->
   
  <!---->
</div>
`;

exports[`EcNavigation should render branding logo and name when given: Branding should have alt and src. 1`] = `
<div
  class="ec-navigation__branding"
>
  <img
    alt="My Branding"
    src="/img/my.png"
  />
</div>
`;
